File: src/handlers/random_release_handler.py
from ask_sdk_core.dispatch_components import AbstractRequestHandler
from ask_sdk_core.utils import is_intent_name
from ask_sdk_core.handler_input import HandlerInput
from ask_sdk_model import Response
from ask_sdk_model.ui import SimpleCard
import logging

from ..utils.discogs_utils import discogs_api

logger = logging.getLogger(__name__)

class RandomReleaseIntentHandler(AbstractRequestHandler):
    """Handler for RandomReleaseIntent"""
    
    def can_handle(self, handler_input: HandlerInput) -> bool:
        return is_intent_name("RandomReleaseIntent")(handler_input)
    
    def handle(self, handler_input: HandlerInput) -> Response:
        logger.info("RandomReleaseIntent triggered")
        
        try:
            release = None
            attempts = 0
            max_attempts = 5
            
            # Try to get a random release, with fallback attempts
            while not release and attempts < max_attempts:
                release = discogs_api.get_random_release()
                attempts += 1
            
            if not release:
                speak_output = "Sorry, I couldn't find a random release right now. Please try again later."
                return (
                    handler_input.response_builder
                    .speak(speak_output)
                    .response
                )
            
            title = release.get('title', 'Unknown Title')
            
            # Get artist names
            artists = release.get('artists', [])
            if artists:
                artist_names = [artist.get('name', 'Unknown Artist') for artist in artists]
                artist_text = ', '.join(artist_names)
            else:
                artist_text = 'Unknown Artist'
            
            year = release.get('year')
            year_text = f" from {year}" if year else ""
            
            genres = release.get('genres', [])
            genre_text = f" in the {', '.join(genres)} genre" if genres else ""
            
            speak_output = f"Here's a random release for you: '{title}' by {artist_text}{year_text}{genre_text}."
            card_content = f"{title}\nby {artist_text}{year_text}{genre_text}"
            
            return (
                handler_input.response_builder
                .speak(speak_output)
                .set_card(SimpleCard("Random Discogs Release", card_content))
                .response
            )
            
        except Exception as e:
            logger.error(f"Error in RandomReleaseIntentHandler: {e}")
            speak_output = "Sorry, I had trouble finding a random release. Please try again later."
            return (
                handler_input.response_builder
                .speak(speak_output)
                .response
            )

File: src/handlers/randomReleaseHandler.js
const Alexa = require('ask-sdk-core');
const discogsAPI = require('../utils/discogsUtils');

const RandomReleaseIntentHandler = {
    canHandle(handlerInput) {
        return Alexa.getRequestType(handlerInput.requestEnvelope) === 'IntentRequest'
            && Alexa.getIntentName(handlerInput.requestEnvelope) === 'RandomReleaseIntent';
    },
    async handle(handlerInput) {
        try {
            console.log('Getting random release...');
            
            const release = await discogsAPI.getRandomRelease();
            const speakOutput = `Here's a random release for you: ${discogsAPI.formatReleaseForSpeech(release)}. 
                ${release.genres && release.genres.length > 0 ? `It's in the ${release.genres.join(' and ')} genre${release.genres.length > 1 ? 's' : ''}.` : ''}`;
            
            const repromptText = 'Would you like another random release or try something else?';

            return handlerInput.responseBuilder
                .speak(speakOutput)
                .reprompt(repromptText)
                .getResponse();

        } catch (error) {
            console.error('Error in RandomReleaseIntentHandler:', error);
            
            const speakOutput = 'Sorry, I had trouble finding a random release. This might be due to a network issue. Please try again later.';
            const repromptText = 'Would you like to try searching for a specific artist or album instead?';

            return handlerInput.responseBuilder
                .speak(speakOutput)
                .reprompt(repromptText)
                .getResponse();
        }
    }
};

module.exports = RandomReleaseIntentHandler;